deps/portal/src/Channel.cpp

```diff
@@ -17,90 +17,143 @@
-    bool Channel::StartInternalThread()
-    {
-        _thread = std::thread(InternalThreadEntryFunc, this);
-        return true;
-    }
-
-    /** Will not return until the internal thread has exited. */
-    void Channel::WaitForInternalThreadToExit()
-    {
-        if (_thread.joinable())
-        {
-            _thread.join();
-        }
-    }
-
-    void Channel::StopInternalThread()
-    {
-        running = false;
-    }
-
-    void Channel::InternalThreadEntry()
-    {
-        while (running)
-        {
-
-            const uint32_t numberOfBytesToAskFor = 65536; // (1 << 16); // This is the value in DarkLighting
-            uint32_t numberOfBytesReceived = 0;
-
-            char buffer[numberOfBytesToAskFor];
-
-            int ret = usbmuxd_recv_timeout(conn, (char *)&buffer, numberOfBytesToAskFor, &numberOfBytesReceived, 100);
-
-            if (ret == 0)
-            {
-                if (numberOfBytesReceived > 0)
-                {
-                    if (running) {
-                        protocol->processData((char *)buffer, numberOfBytesReceived);
-                    }
-                }
-            }
-            else
-            {
-                portal_log("%d: There was an error receiving data\n", conn);
-                running = false;
-            }
-        }
-    }
-
-    void Channel::simpleDataPacketProtocolDelegateDidProcessPacket(std::vector<char> packet, int type, int tag)
-    {
-        std::shared_ptr<ChannelDelegate> strongDelegate = delegate.lock();
-        if (strongDelegate) {
-            strongDelegate->channelDidReceivePacket(packet, type, tag);
-        }
-    }
+	uint32_t sentBytes = 0;
+
+	usbmuxd_send(conn, &data[0], data.size(), &sentBytes);
+}
+
+void Channel::setState(State state)
+{
+	if (state == getState()) {
+		return;
+	}
+
+	//std::cout << "Channel:setState: " << state << std::endl;
+	_state = state;
+
+	if (auto delegate = this->delegate.lock()) {
+		delegate->channelDidChangeState(state);
+	}
+}
+
+/** Returns true if the thread was successfully started, false if there was an error starting the thread */
+bool Channel::StartInternalThread()
+{
+	_thread = std::thread(InternalThreadEntryFunc, this);
+	return true;
+}
+
+/** Will not return until the internal thread has exited. */
+void Channel::WaitForInternalThreadToExit()
+{
+	//if (_thread == nullptr) {
+	//	return;
+	//}
+	std::unique_lock<std::mutex> lock(worker_mutex);
+
+	if (_thread.joinable() == true) {
+		_thread.join();
+	}
+
+	//_thread = nullptr;
+	lock.unlock();
+}
+
+void Channel::StopInternalThread()
+{
+	running = false;
+}
+
+// TODO: https://stackoverflow.com/questions/58477291/function-exceeds-stack-size-consider-moving-some-data-to-heap-c6262
+void Channel::InternalThreadEntry()
+{
+	while (running) {
+
+		const uint32_t numberOfBytesToAskFor =
+			65536; // (1 << 16); // This is the value in DarkLighting
```

Comment:
How big is the biggest message being sent? What is the biggest video frame that could be sent at 60mbits?

Reply:
It's hard to know for sure, as it's a compressed h264 frame. I'm not sure of a reliable way to estimate the number of bytes each frame will be.

Reply:
It could be helpful to add some logging and just print out what the max size is you see for a session and use that.

My thinking is: you want to do a single read that will get at least one entire frame.

Reply:
So a smaller read might end up being more performant, as we're not waiting for more than one frame to be buffered before processing it? Makes sense!

Reply:
The way that the OS deals with big reads is that it will return as soon as there is _something_ to return, so it isn't true that smaller reads will be better.

Reply:
They range quite a bit.  `10` to `479539` with most being around ~`223920`

See the dump [here](https://gist.github.com/wtsnz/09b351c7aebe7063662bd899eae77379)

Reply:
Interesting. I wonder how things would perform if you used `1<<18` (262144) so that it's more than the average.